src/Domain/ElementDistribution.cpp

```diff
@@ -0,0 +1,165 @@
+        gsl::at(element_id.segment_ids(), i).refinement_level(), i);
+  }
+  alg::sort(dimension_by_highest_refinement_level,
+            [](const std::pair<size_t, size_t>& lhs,
+               const std::pair<size_t, size_t>& rhs) {
+              return lhs.first < rhs.first;
+            });
+
+  size_t element_order_index = 0;
+
+  // 'gap' the lowest refinement direction bits as:
+  // ... x1 x0 -> ... x1 0 0 x0,
+  // then bitwise or in 'gap'ed and shifted next-lowest refinement direction
+  // bits as:
+  // ... y2 y1 y0 -> ... y2 0 y1 x1 0 y0 x0
+  // then bitwise or in 'gap'ed and shifted highest-refinement direction bits
+  // as:
+  // ... z3 z2 z1 z0 -> z3 z2 y2 z1 y1 x1 z0 y0 x0
+  // note that we must skip refinement-level 0 dimensions as though they are
+  // not present
+  size_t leading_gap = 0;
+  for (size_t i = 0; i < Dim; ++i) {
+    const size_t id_to_gap_and_shift =
+        gsl::at(element_id.segment_ids(),
+                gsl::at(dimension_by_highest_refinement_level, i).second)
+            .index();
+    size_t total_gap = leading_gap;
+    if (gsl::at(dimension_by_highest_refinement_level, i).first > 0) {
+      ++leading_gap;
+    }
+    for (size_t bit_index = 0;
+         bit_index < gsl::at(dimension_by_highest_refinement_level, i).first;
+         ++bit_index) {
+      element_order_index |=
+          ((id_to_gap_and_shift & two_to_the(bit_index)) << total_gap);
+      for (size_t j = 0; j < Dim; ++j) {
+        if (i != j and
+            bit_index + 1 <
+                gsl::at(dimension_by_highest_refinement_level, j).first) {
+          ++total_gap;
+        }
+      }
+    }
+  }
+  return element_order_index;
+}
+}  // namespace
+
+template <size_t Dim>
+BlockZCurveProcDistribution<Dim>::BlockZCurveProcDistribution(
+    size_t number_of_procs,
+    const std::vector<std::array<size_t, Dim>>& refinements_by_block) noexcept {
+  elements_per_proc_.reserve(number_of_procs);
+  block_element_distribution_.reserve(refinements_by_block.size());
+  auto add_number_of_elements_for_refinement =
+      [](size_t lhs, const std::array<size_t, Dim>& rhs) {
+        size_t value = 1;
+        for (size_t i = 0; i < Dim; ++i) {
+          value *= two_to_the(gsl::at(rhs, i));
+        }
+        return lhs + value;
+      };
+  const size_t number_of_elements =
+      std::accumulate(refinements_by_block.begin(), refinements_by_block.end(),
+                      0_st, add_number_of_elements_for_refinement);
+  ASSERT(not refinements_by_block.empty(),
+         "`refinements_by_block` must be non-empty.");
+  // currently, we just assign uniform weight to elements. In future, it will
+  // probably be better to take into account p-refinement, but then the z-curve
+  // method will also require weighting.
+  size_t remaining_elements_in_block =
+      add_number_of_elements_for_refinement(0_st, refinements_by_block[0]);
+  size_t current_block = 0;
+  block_element_distribution_.emplace_back();
```

Comment:
For my understanding: why are you using `reserve()` (above) then `emplace_back()` (here and below)... vs assigning the vector above with a statement like `block_element_distribution_ =  std::vector<std::vector<std::pair<size_t, size_t>>>(size)`?

I think the logic would be simpler without interleaving the construction of default elements, but am I missing something?

Reply:
I agree -- this works by just creating a vector of the right size from the start -- I'll do that.